host/xtest/regression_1000.c

```diff
@@ -2349,3 +2349,214 @@ static void xtest_tee_test_1032(ADBG_Case_t *c)
 }
 ADBG_CASE_DEFINE(regression, 1032, xtest_tee_test_1032,
 		"Register read-only shared memory");
+
+/*
+ * regression_1033
+ * Depends on kernel commit "tee: add support for application-based session login methods"
+ * Mailing list archives: https://lists.trustedfirmware.org/pipermail/op-tee/2020-October/000211.html
+ * Upstream: <put sha-1 here when known>
+ *
+ * xtest skips the test when not built with OpenSSL.
+ */
+static void xtest_tee_test_1033(ADBG_Case_t *c)
+{
+#ifdef OPENSSL_FOUND
+	TEEC_Result result = TEEC_ERROR_GENERIC;
+	uint32_t ret_orig = 0;
+	TEEC_Session session = { };
+	uint32_t login = UINT32_MAX;
+	TEEC_UUID client_uuid = { };
+	TEEC_UUID expected_client_uuid = { };
+	TEEC_UUID uuid_ns = { };
+	char uuid_name[TEE_UUID_NS_NAME_SIZE + 4] = { };
+	char app[PATH_MAX] = { };
+
+	result = xtest_uuid_from_str(&uuid_ns, client_uuid_linux_ns);
+
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c, result))
+		return;
+
+	if (readlink("/proc/self/exe", app, PATH_MAX) < 0) {
+		Do_ADBG_Log(" - 1033 -   skip test, "
+			    "/proc/self/exe not readable");
+		return;
+	}
+
+	sprintf(uuid_name, "app=%s", app);
+	uuid_name[TEE_UUID_NS_NAME_SIZE] = 0;
```

Comment:
This isn't needed with `snprintf()` and if it's needed with `sprintf()` you're already in deep trouble.

Reply:
The kernel buffer is `PATH_MAX` long, because having a buffer longer than that is more problematic to allocate. But since a path with added `app=` and `uid=` and stuff can be longer, the kernel truncates the buffer to `PATH_MAX`. This is considered safe since the TA checks whether the hash is an expected one or not, and the TA authors should not allow for collision attacks based on the application ID being truncated.

The kernel implementation in fact uses `snprintf()` for this. However, this is not possible in optee_test, since GCC complains about the string being truncated:
```
regression_1000.c: In function ‘xtest_tee_test_1033’:
regression_1000.c:2381:50: error: ‘%s’ directive output may be truncated writing up to 4095 bytes into a region of size 4092 [-Werror=format-truncation=]
  snprintf(uuid_name, TEE_UUID_NS_NAME_SIZE, "app=%s", app);
                                                  ^~   ~~~
regression_1000.c:2381:2: note: ‘snprintf’ output between 5 and 4100 bytes into a destination of size 4096
  snprintf(uuid_name, TEE_UUID_NS_NAME_SIZE, "app=%s", app);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
regression_1000.c: In function ‘xtest_tee_test_1034’:
regression_1000.c:2449:57: error: ‘%s’ directive output may be truncated writing up to 4095 bytes into a region of size between 4079 and 4086 [-Werror=format-truncation=]
  snprintf(uuid_name, TEE_UUID_NS_NAME_SIZE, "uid=%x:app=%s", geteuid(), app);
                                                         ^~              ~~~
regression_1000.c:2449:2: note: ‘snprintf’ output between 11 and 4113 bytes into a destination of size 4096
  snprintf(uuid_name, TEE_UUID_NS_NAME_SIZE, "uid=%x:app=%s", geteuid(), app);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
regression_1000.c: In function ‘xtest_tee_test_1035’:
regression_1000.c:2520:57: error: ‘%s’ directive output may be truncated writing up to 4095 bytes into a region of size between 4079 and 4086 [-Werror=format-truncation=]
  snprintf(uuid_name, TEE_UUID_NS_NAME_SIZE, "gid=%x:app=%s", group, app);
                                                         ^~          ~~~
regression_1000.c:2520:2: note: ‘snprintf’ output between 11 and 4113 bytes into a destination of size 4096
  snprintf(uuid_name, TEE_UUID_NS_NAME_SIZE, "gid=%x:app=%s", group, app);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Hence the truncation is performed manually here. Though I will replace `sprintf()` with `snprintf()` anyway for hardening purposes, but with an actual buffer size.

Reply:
Something like (might need some tweaking in the formatting)
``` C
snprintf(uuid_name, sizeof(uuid_name), "app=%*s", TEE_UUID_NS_NAME_SIZE, app);
```
should do it.

Reply:
`"app=%.*s"` works.

But then the next parameter should supply not `TEE_UUID_NS_NAME_SIZE`, but `TEE_UUID_NS_NAME_SIZE - x`, where `x` is the amount of other characters in the string ("app="), since this parameter limits the length of the `%s` only. 

While this may work in the case of "app=%s", this wont in case of "uid=XXX:app=", since we don't know the length of the printed uid without calculating it. And at this point, manually setting a byte to `\0` is much easier.

Reply:
OK

Reply:
I just commented on the kernel patch about the length of uuid_name.

So if we try to get rid of the limit altogether then we don't need to care about it anymore.